#Homework assignment #3

These problem sets focus on using the Beautiful Soup library to scrape web pages.

##Problem Set #1: Basic scraping

I've made a web page for you to scrape. It's available [here](http://static.decontextualize.com/widgets2015.html). The page concerns the catalog of a famous [widget](http://en.wikipedia.org/wiki/Widget) company. You'll be answering several questions about this web page. First off, in the cell below, write some code so that you end up with a variable called `html_str` that contains the HTML source code of the page.

In [23]:
import urllib
html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()

Now, in the cell below, use Beautiful Soup to write an expression that evaluates to the number of `<h3>` tags contained in `widgets.html`. I've added the Beautiful Soup import statement for you.

In [24]:
from bs4 import BeautifulSoup
document = BeautifulSoup(html_str)
document

/Library/Python/2.7/site-packages/bs4/__init__.py:166: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html.parser"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

To get rid of this warning, change this:

 BeautifulSoup([your markup])

to this:

 BeautifulSoup([your markup], "html.parser")

  markup_type=markup_type))


<!DOCTYPE doctype html>\n\n<html>\n<head>\n<title>Fearless Widget Factory Catalog</title>\n<style type="text/css">\n\t\t\ttable.widgetlist {\n\t\t\t\tborder-collapse: collapse;\n\t\t\t\tborder: 1px #333 solid;\n\t\t\t\twidth: 50%;\n\t\t\t}\n\t\t\ttable.widgetlist td {\n\t\t\t\tborder: 1px #333 solid;\n\t\t\t\tpadding: 5px;\n\t\t\t}\n\t\t\ttable.widgetlist thead {\n\t\t\t\tbackground: #333;\n\t\t\t\tcolor: white;\n\t\t\t}\n\t\t</style>\n</head>\n<body>\n<h1>Fearless Widget Factory</h1>\n<p><i>Purveyor of fearless widgets since 1956</i></p>\n<h2>Widget Catalog</h2>\n<p>Please call <span class="tel">212-555-9912</span> for latest prices.</p>\n<h3>Forensic Widgets</h3>\n<table class="widgetlist">\n<thead>\n<td>Part No.</td>\n<td>Name</td>\n<td>Price</td>\n<td>Quantity in warehouse</td>\n</thead>\n<tr class="winfo">\n<td class="partno">C1-9476</td>\n<td class="wname"><b>Skinner Widget</b></td>\n<td class="price">$2.70</td>\n<td class="quantity">512</td>\n</tr>\n<tr class="winfo">\n<td class

In [25]:
h3_tag = document.find_all('h3')
h3_list = [h3.string for h3 in h3_tag]
#print h3_list
len (h3_list)

3

Now, in the cell below, write an expression or series of statements that displays the telephone number beneath the "Widget Catalog" header.

In [26]:
telphone_number = document.find('span', attrs={'class': 'tel'})
telphone_number.string

u'212-555-9912'

In the cell below, use Beautiful Soup to write some code that creates a list of the names of all the widgets on the page. After your code has executed, `widget_names` should evaluate to a list that looks like this (though not necessarily in this order):

    [u'Skinner Widget',
     u'Widget For Furtiveness',
     u'Widget For Strawman',
     u'Manicurist Widget',
     u'Infinite Widget',
     u'Yellow-Tipped Widget',
     u'Unshakable Widget',
     u'Self-Knowledge Widget',
     u'Widget For Cinema']

In [27]:
b_tag = document.find_all('b')
[b.string for b in b_tag]

[u'Skinner Widget',
 u'Widget For Furtiveness',
 u'Widget For Strawman',
 u'Manicurist Widget',
 u'Infinite Widget',
 u'Yellow-Tipped Widget',
 u'Unshakable Widget',
 u'Self-Knowledge Widget',
 u'Widget For Cinema']

##Problem set #2: Widget dictionaries

For this problem set, we'll continue to use the HTML page from the previous problem set. In the cell below, I've made an empty list and assigned it to a variable called `widgets`. Write code that populates this list with dictionaries, one dictionary per widget in the source file. The keys of each dictionary should be `partno`, `widgetname`, `price`, and `quantity`, and the value for each of the keys should be the value for the corresponding column for each row. After executing the cell, your list should look something like this:

    [{'partno': u'C1-9476',
      'price': u'$2.70',
      'quantity': u'512',
      'wname': u'Skinner Widget'},
     {'partno': u'JDJ-32/V',
      'price': u'$9.36',
      'quantity': u'967',
      'wname': u'Widget For Furtiveness'},
      ...several items omitted...
     {'partno': u'5B-941/F',
      'price': u'$13.26',
      'quantity': u'919',
      'wname': u'Widget For Cinema'}]

And this expression:

    widgets[5]['partno']
    
... should evaluate to:

    u'MZ-556/B'
    
(Hint: This could get a little messy! Each column in the table might need to be treated a bit differently to get the appropriate strings out.)

In [28]:
widgets = []

html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
document = BeautifulSoup(html_str)
#print document

tr_tags = document.find_all('tr')
for tr_tag in tr_tags:
    td_tags = tr_tag.find_all('td')
    widgets_dic = {}
    widgets_dic['partno'] = td_tags[0].string
    widgets_dic['price'] = td_tags[2].string
    widgets_dic['quantity'] = td_tags[3].string
    widgets_dic['wname'] = td_tags[1].string
    widgets.append(widgets_dic)

#widgets

widgets[5]['partno']

u'MZ-556/B'

In the cell below, duplicate your code from the previous question. Modify the code to ensure that the values for `price` and `quantity` in each dictionary are floating-point numbers and integers, respectively. I.e., after executing the cell, your code should display something like this:

    [{'partno': 'C1-9476',
      'price': 2.7,
      'quantity': 512,
      'widgetname': 'Skinner Widget'},
     {'partno': 'JDJ-32/V',
      'price': 9.36,
      'quantity': 967,
      'widgetname': 'Widget For Furtiveness'},
     ... some items omitted ...
     {'partno': '5B-941/F',
      'price': 13.26,
      'quantity': 919,
      'widgetname': 'Widget For Cinema'}]

(Hint: Use the `float()` and `int()` functions. You may need to use string slices to convert the `price` field to a floating-point number.)

In [29]:
widgets = []

html_str = urllib.urlopen("http://static.decontextualize.com/widgets2015.html").read()
document = BeautifulSoup(html_str)

tr_tags = document.find_all('tr')
for tr_tag in tr_tags:
    td_tags = tr_tag.find_all('td')
    widgets_dic = {}
    widgets_dic['partno'] = td_tags[0].string
    widgets_dic['price'] = float(td_tags[2].string.split('$')[1])
    widgets_dic['quantity'] = int(td_tags[3].string)
    widgets_dic['wname'] = td_tags[1].string
    widgets.append(widgets_dic)


#for i in len(widgets):
widgets

[{'partno': u'C1-9476',
  'price': 2.7,
  'quantity': 512,
  'wname': u'Skinner Widget'},
 {'partno': u'JDJ-32/V',
  'price': 9.36,
  'quantity': 967,
  'wname': u'Widget For Furtiveness'},
 {'partno': u'YP4-325/J',
  'price': 5.17,
  'quantity': 787,
  'wname': u'Widget For Strawman'},
 {'partno': u'VK-486',
  'price': 8.97,
  'quantity': 441,
  'wname': u'Manicurist Widget'},
 {'partno': u'R4K-990',
  'price': 11.73,
  'quantity': 320,
  'wname': u'Infinite Widget'},
 {'partno': u'MZ-556/B',
  'price': 2.35,
  'quantity': 948,
  'wname': u'Yellow-Tipped Widget'},
 {'partno': u'QV-730',
  'price': 3.76,
  'quantity': 59,
  'wname': u'Unshakable Widget'},
 {'partno': u'T1-9731',
  'price': 7.11,
  'quantity': 790,
  'wname': u'Self-Knowledge Widget'},
 {'partno': u'5B-941/F',
  'price': 13.26,
  'quantity': 919,
  'wname': u'Widget For Cinema'}]

Great! I hope you're having fun. In the cell below, write an expression that uses the `widgets` list created in the cell above to calculate the total number of widgets that the factory has in its warehouse.

Expected output: `5743`

In [30]:
total= []
for i in range(len(widgets)):
    total.append(widgets[i]['quantity'])
sum(total)

5743

In the cell below, write an expression (or series of expressions) that evaluates to a list the names of widgets containing only those widgets whose price is above $9.30.

Expected output: `[u'Widget For Furtiveness', u'Infinite Widget', u'Widget For Cinema']`

In [31]:
list = []
for i in range(len(widgets)):
    if widgets[i]['price']>9.30:
        list.append(widgets[i]['wname'])
print list

[u'Widget For Furtiveness', u'Infinite Widget', u'Widget For Cinema']


##Problem set #3: Sibling rivalries

In the following problem set, you will yet again be working with the data in `widgets.html`. In order to accomplish the tasks in this problem set, you'll need to learn about Beautiful Soup's `.find_next_sibling()` method. Here's some information about that method, cribbed from the notes:

Often, the tags we're looking for don't have a distinguishing characteristic, like a class attribute, that allows us to find them using `.find()` and `.find_all()`, and the tags also aren't in a parent-child relationship. This can be tricky! For example, take the following HTML snippet, (which I've assigned to a string called `example_html`):

In [32]:
example_html = """
<h2>Camembert</h2>
<p>A soft cheese made in the Camembert region of France.</p>

<h2>Cheddar</h2>
<p>A yellow cheese made in the Cheddar region of... France, probably, idk whatevs.</p>
"""

If our task was to create a dictionary that maps the name of the cheese to the description that follows in the `<p>` tag directly afterward, we'd be out of luck. Fortunately, Beautiful Soup has a `.find_next_sibling()` method, which allows us to search for the next tag that is a sibling of the tag you're calling it on (i.e., the two tags share a parent), that also matches particular criteria. So, for example, to accomplish the task outlined above:

In [33]:
example_doc = BeautifulSoup(example_html)
cheese_dict = {}
for h2_tag in example_doc.find_all('h2'):
    cheese_name = h2_tag.string
    cheese_desc_tag = h2_tag.find_next_sibling('p')
    cheese_dict[cheese_name] = cheese_desc_tag.string

cheese_dict

{u'Camembert': u'A soft cheese made in the Camembert region of France.',
 u'Cheddar': u'A yellow cheese made in the Cheddar region of... France, probably, idk whatevs.'}

With that knowledge in mind, let's go back to our widgets. In the cell below, write code that uses Beautiful Soup, and in particular the `.find_next_sibling()` method, to find out how many widgets are in the table *just beneath* the header "Hallowed Widgets." (You can tell by looking at the page that there are four such widgets. But this is a programming class, so we have to write a program to do it.)

In [34]:
h3_tags = document.find_all('h3')
for h3_tag in h3_tags: 
    if h3_tag.string == "Hallowed widgets":
        table_tags = h3.find_next_sibling('table')
        #print table_tags
        item_tags = table_tags.find_all('tr')
        #print item_tags
print len(item_tags)

4


Okay, now, the final task. If you can accomplish this, you are truly an expert web scraper. I'll have little web scraper certificates made up and I'll give you one, if you manage to do this thing. And I know you can do it!

In the cell below, I've created a variable `category_counts` and assigned to it an empty dictionary. Write code to populate this dictionary so that its keys are "categories" of widgets (e.g., the contents of the `<h3>` tags on the page: "Forensic Widgets", "Mood widgets", "Hallowed Widgets") and the value for each key is the number of widgets that occur in that category. I.e., after your code has been executed, the dictionary `category_counts` should look like this:

    {u'Forensic Widgets': 3, u'Hallowed widgets': 4, u'Mood widgets': 2}


In [35]:
category_counts  = {}
h3_tags = document.find_all('h3')

for h3_tag in h3_tags: 
    if h3_tag.string == "Hallowed widgets":
        table_tags = h3_tag.find_next_sibling('table')
        item_tags = table_tags.find_all('tr')
        category_counts[u'Hallowed widgets']=len(item_tags)
    elif h3_tag.string == "Forensic Widgets":
        table_tags = h3_tag.find_next_sibling('table')
        item_tags = table_tags.find_all('tr')
        category_counts[u'Forensic Widgets']=len(item_tags) 
    elif h3_tag.string == "Mood widgets":
        table_tags = h3_tag.find_next_sibling('table')
        item_tags = table_tags.find_all('tr')
        category_counts[u'Mood widgets']=len(item_tags)
print category_counts

{u'Forensic Widgets': 3, u'Mood widgets': 2, u'Hallowed widgets': 4}


Congratulations! You're done.